online-shop/src/app/data.service.ts

```diff
@@ -0,0 +1,142 @@
+import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable, throwError } from 'rxjs';
+import { IProduct } from './product-list/product-list.component';
+import { catchError } from 'rxjs/operators';
+
+export interface ICartProduct {
+  productId: number,
+  category: string,
+  name: string,
+  price: number,
+  quantity: number,
+}
+
+interface IOrder {
+  customer: string,
+  products: ICartProduct[],
+}
+
+export interface IUser {
+  username: string,
+  password: string | null,
+  fullName?: string,
+  roles?: string[]
+}
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class DataService {
+  private shoppingCart: ICartProduct[] = [];
+  private BASE_URL = "http://localhost:3000";
+  private currentUser: IUser = { username: "", password: null };
+  private isCustomer: boolean = false;
+  private isAdmin: boolean = false;
+
+
+  constructor(private http: HttpClient) { }
+
+  private handleError(error: HttpErrorResponse) {
```

Comment:
Add return type, even if it's void.
- improve the clarity of the code: you do not have to read the whole function body to see if it returns anything
- ensure type-safe: you will never assign the function with the void return type to a variable